platforms/core-configuration/file-operations/src/main/java/org/gradle/api/internal/file/DefaultProjectLayout.java

```diff
@@ -33,9 +33,12 @@
 import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.util.PatternSet;
 import org.gradle.internal.Factory;
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
 
 import java.io.File;
 
+@ServiceScope(Scope.Project.class)
 public class DefaultProjectLayout implements ProjectLayout, TaskFileVarFactory {
```

Comment:
Could this be dropped if we registered `ProjectLayout` as the service?

Reply:
We already register it as a service. The current reason is due to `DefaultProject` injecting the implementation, rather than the interfaces:
https://github.com/gradle/gradle/blob/e05529fa9fac0f324ca198425c4596130ef76a39/subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java#L1001

It seems, this is only due to the need to do `DefaultProjectLayout#setBuildDirectory`, which according to the comment should go away:
https://github.com/gradle/gradle/blob/fb8cc203942d2ac5f5d1fca2bd40179981a8fd0c/platforms/core-configuration/file-operations/src/main/java/org/gradle/api/internal/file/DefaultProjectLayout.java#L126-L131

It's a nice catch, I'll fix it in a follow-up to avoid mixing non-trivial changes to this PR.

Reply:
Here is the follow-up. It resolved your question, but also uncovered more questions about the buildDirectory setter
https://github.com/gradle/gradle/pull/32655